packages/story-editor/src/app/rightClickMenu/provider.js

```diff
@@ -61,14 +39,14 @@ import rightClickMenuReducer, {
   ACTION_TYPES,
   DEFAULT_RIGHT_CLICK_MENU_STATE,
 } from './reducer';
-import { getElementStyles } from './utils';
 import useLayerSelect from './useLayerSelect';
-
-const UNDO_HELP_TEXT = sprintf(
-  /* translators: %s: Ctrl/Cmd + Z keyboard shortcut */
-  __('Press %s to undo the last change', 'web-stories'),
-  prettifyShortcut('mod+z')
-);
+import {
+  useCopyPasteActions,
+  useElementActions,
+  useLayerActions,
+  usePageActions,
+  usePresetActions,
+} from './hooks';
```

Comment:
I'll have more time to review it next week but generally makes sense to me, dividing into hooks like this.

We could additionally even create a separate file for each element as well so that the relevant hooks would be called in those files and the provider would just use the element hooks.
E.g. for text
```
// In provider.
const textItems = useTextItems(...);
const pageItems = usePageItems(...);

// In `useTextItems`
import { useCopyPasteActions, useElementActions, useLayerActions } from './hooks`;
...
// In `usePageItems`
import { useCopyPasteActions, usePageActions, useLayerActions } from './hooks';

... etc.
```

In this case, if there are items inside the `useElementActions` that are for text only, those could be added to `useTextItems` directly and `useElementActions` could include actions that are for at least 2 element types (unique items would be in the hook per element instead).

Just an idea without having gone through all the hook files to see how many unique/common items there are. Thoughts?

Reply:
Thanks for looking at this! Sorry it's so big. I prefer to make smaller prs, but since this mostly is moving code around I figured it'd be better to do it all at once

I do like the idea of having hooks for each element type and grabbing those options. However, I've had experience with nesting custom hooks inside other custom hooks. IMO it's better not to nest custom hooks inside other custom hooks because the logic becomes very spread out and dependencies can get muddy. One change in a hook can affect `useEffects` in other hooks and it's hard to tell until a bug comes up.

I prefer to keep hooks one level deep. What do you think?

I debated making just hooks for each menu but there would be a lot of repeated code in the actions for each. However, we could make pure functions for the actions and pass in the necessary scoped variables and functions as arguments.

Reply:
➕ Yea I'm in favor of minimizing nesting of hooks. Adds complexity and scope to tracking down causes of a re-render

Reply:
💭 A lot of these `menuItems` mirror composition in jsx...

Instead of using switch statements and all this object composition in the provider, could we just achieve all this composition in the markup? so in `RightClickMenu` it might just look something like this:
```
const menuType = deriveMenuType(selectedElements);
const menu = useMemo(() => {
  switch(menuType) {
    case MENU_TYPES.MULTI_ELEMENT:
      return <MultiElementMenu />;
    case MENU_TYPES.SHAPE:
      return <ShapeMenu />
    ...
  }
}, [menuType]);
...
render createPortal(
  ...
  {menu}
  ...
);
```
then we can compose the menus through jsx in the individual menus:
```
// ShapeMenu.js
function ShapeMenu() {
  const presetActions = usePresetActions();
  ...
  return (
    <>
      <ContextMenuComponents.MenuItem 
        label={RIGHT_CLICK_MENU_LABELS.DUPLICATE_ELEMENTS(
          selectedElements.length
        )}
        onClick={elementActions.onDuplicateSelectedElements},
        separator="bottom"
      />
      ...
      <LayerSubMenu />
    </>
  );
}
```

I think this would have the benefit of:
- not needing to create every action and every menu on every render
- drastically simplifying the right click menu provider
- remove some needless abstraction
- leaving the composition for the part of react that's really good at composition, the jsx

Reply:
I like this idea a lot, Max